feat: LitRenderer

tomivirkki opened the pull request:
Closes #856

tomivirkki committed:
feat: LitRenderer

tomivirkki committed:
test: add LitRenderer tests

tomivirkki committed:
chore: cleanup

knoobie commented on vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridLitRendererPage.java:
```diff
@@ -0,0 +1,73 @@
+        grid.addColumn(LitRenderer
+                .<Integer> of(
+                        "<span id=\"item-${index}\">Lit: ${item.name}</span>")
+                .withProperty("name", item -> "Item " + item));
```
Can I ask for a test, that verifies that using two `LitRenderer` with the same `property` name can be used in two seperat columns without overriding the previous set `property`? :) Xref: https://github.com/vaadin/flow-components/issues/916#issuecomment-832493018

tomivirkki replied:
👍 Thanks for the suggestion. Will add the test.

tomivirkki replied:
Covered in https://github.com/vaadin/flow-components/pull/1838/commits/048d4376c7ac21b9db9d5688272dd6399a1d2271

tomivirkki committed:
fix: support renderer instance specific property names

tomivirkki committed:
test: remove old data generator registrations

tomivirkki committed:
refactor: cleanup property namespace

tomivirkki committed:
refactor: support reattaching the host component

tomivirkki committed:
refactor: define a Renderer interface (Grid/Grid Pro wip)

tomivirkki committed:
refactor: revert back to extending Renderer

tomivirkki committed:
chore: fix formatter issues

tomivirkki committed:
docs: document LitRenderer class

tomivirkki committed:
test: exclude devserver from serializable test

tomivirkki committed:
feat: drop support for getEventHandlers, add getFunction

tomivirkki committed:
test: exclude devserver from serializable test

tomivirkki committed:
chore: cleanup pom.xml

tomivirkki committed:
test: stabilize tests

tomivirkki committed:
docs: document lit-renderer.ts

tomivirkki committed:
docs: readme

tomivirkki commented on flow-renderer-parent/flow-renderer/src/main/resources/META-INF/resources/frontend/lit-renderer.ts:
```diff
@@ -0,0 +1,108 @@
+  rendererName: string,
+  rendererId: string
+) => {
+  if (component[rendererName]?.__rendererId === rendererId) {
```
The check for `__rendererId` property is necessary since the `renderer` function may get overridden by another renderer, for example, by one coming from `vaadin-template-renderer`. We don't want LitRenderer registration cleanup to unintentionally remove the new renderer.

tomivirkki commented on flow-renderer-parent/flow-renderer/src/test/java/com/vaadin/flow/renderer/tests/RendererSerializableTest.java:
```diff
@@ -0,0 +1,15 @@
+package com.vaadin.flow.renderer.tests;
```
These I needed to add to make CI pass. See https://github.com/vaadin/flow/issues/11371 Not related to this feature in any way.

tomivirkki replied:
Resolved by https://github.com/vaadin/flow/issues/11371#issuecomment-872774648

tomivirkki commented on vaadin-virtual-list-flow-parent/vaadin-virtual-list-flow/src/main/resources/META-INF/resources/frontend/virtualListConnector.js:
```diff
@@ -49,31 +49,41 @@ window.Vaadin.Flow.virtualListConnector = {
-        originalRenderer.call(list, root, list, model);
+    // Add an observer function that will invoke on virtualList.renderer property
+    // change and then patches it with a wrapper renderer
+    list.patchVirtualListRenderer = function () {
```
Needed to change this logic a bit.

The previous revision assumed the `virtual-list` already has a `renderer` function at this point and patched it. That may not always be the case. For example, a `LitRenderer` or a `<template>` tag may be added dynamically. In that case, the connector would end up patching a non-existing function and also the patched function could get overridden later.

The new version observes for changes in `virtual-list.renderer` and patches the function when it's assigned.

tomivirkki commented on vaadin-virtual-list-flow-parent/vaadin-virtual-list-flow/src/main/java/com/vaadin/flow/component/virtuallist/VirtualList.java:
```diff
@@ -209,16 +201,33 @@ public void setRenderer(ValueProvider<T, String> valueProvider) {
+        rendererRegistrations.clear();
+
+        Rendering<T> rendering;
+        if (renderer instanceof LitRenderer) {
+            // LitRenderer
+            if (template.getParent() != null) {
+                getElement().removeChild(template);
+            }
+            rendering = renderer.render(getElement(),
+                    dataCommunicator.getKeyMapper());
+        } else {
+            // TemplateRenderer or ComponentRenderer
+            if (template.getParent() == null) {
+                getElement().appendChild(template);
+            }
+            rendering = renderer.render(getElement(),
+                    dataCommunicator.getKeyMapper(), template);
```
A `<template>` and a `renderer` function (one not assigned by `vaadin-template-renderer`) can't exist on a component at the same time.

This logic removes the `<template>` element from the host in case a `LitRenderer`'s `renderer` function is to be assigned. And respectively, restores the `<template>` element when a `TemplateRenderer` or a `ComponentRenderer` is set.

tomivirkki commented on flow-renderer-parent/flow-renderer/src/main/java/com/vaadin/flow/renderer/LitRenderer.java:
```diff
@@ -0,0 +1,423 @@
+     * @see <a href=
+     *      "https://lit.dev/docs/templates/expressions/#event-listener-expressions">https://lit.dev/docs/templates/expressions/#event-listener-expressions</a>
+     */
+    public LitRenderer<T> withFunction(String functionName,
```
Naming TBD: withFunction, withEventHandler, withClientCallable....?

tomivirkki committed:
chore: fix typo

tomivirkki committed:
chore: Fix formatter issue

tomivirkki committed:
test: revert exclude devserver from serializable test

manolo commented on pom.xml:
```diff
@@ -37,6 +37,7 @@
         <osgi.bundle.license>http://www.apache.org/licenses/LICENSE-2.0</osgi.bundle.license>
     </properties>
     <modules>
+        <module>flow-renderer-parent</module>
```
let's align with other modules
`vaadin-renderer-flow-parent` 
`vaadin-renderer-flow`
`vaadin-renderer-flow-integration-tests`

tomivirkki replied:
Done

tomivirkki committed:
refactor: change module names

ZheSun88 commented on vaadin-renderer-flow-parent/pom.xml:
```diff
@@ -0,0 +1,30 @@
+    <name>Flow Renderer Parent</name>
+    <modules>
+        <module>vaadin-renderer-flow</module>
+        <module>vaadin-renderer-flow-integration-tests</module>
```
```
<profile>
            <id>default</id>
            <activation>
                <property>
                    <name>!release</name>
                </property>
            </activation>
            <modules>
                <module>vaadin-renderer-flow-integration-tests</module>
            </modules>
        </profile>
  ```
this is not correct.  this module should be removed from here, as under there is a profile to add it to the maven reactor. 

i think that is the reason why the build failed for this PR.

tomivirkki replied:
Thanks, that did it. Now the tests passed

ZheSun88 replied:
great

tomivirkki committed:
fix: update pom.xml

Commit:
chore: comment why module name is `vaadin-renderer-flow`

Commit:
fix: a typo in argument name in LitRenderer.prepare()

Commit:
chore: include a comment from PR review to the source code

If it was worth mentioning in the PR review, it may as well be just added to the source code.

Commit:
chore: fix a few Sonar- and IDEA-reported code issues

Commit:
chore: fix code formatting

Commit:
chore: extracted lambda into a separate `ComboBox.render()` method

Commit:
fix: JavaDoc comments for the LitRenderer / LitRendering APIs

Also renamed `LitRendering.getRendererRegistration()` to `LitRendering.getRenderingRegistration()` since the registration is rendering-specific.

Commit:
chore: remove nesting of data generators in LitRendering

Remove unnecessary layer of composite / nesting of the data generators contained in a LitRendering.

All data generators added to a `LitRendering` based on `withProperty()` calls are combined into a single `CompositeDataGenerator`. This composite data generator was in turn wrapped into another `CompositeDataGenerator` before making it available via `LitRendering.getDataGenerator()`. This double wrapping made it easy to clean up the data generators when the rendering is disposed even if the user of the rendering did not clean up the rendering's data generator explicitly (because after the rendering is disposed, the rendering's data generator becomes empty).

However, it's recommended to remove the rendering's data generator from a component after the rendering is disposed. And so the double nesting does not seem to be necessary, and is now removed.

Commit:
Merge branch 'master' into lit-renderer

vlukashov commented:
Looks good to me. I think it could be merged as soon as `master` becomes Vaadin 22.

vursen commented on vaadin-renderer-flow-parent/vaadin-renderer-flow/src/main/resources/META-INF/resources/frontend/lit-renderer.ts:
```diff
@@ -0,0 +1,112 @@
+    // TODO: Remove once https://github.com/vaadin/web-components/issues/2161 is done
+    if (root.__litRenderer !== renderer) {
+      root.innerHTML = '';
+      delete root._$litPart$;
```
Removing `_$litPart$` may be no longer necessary here since https://github.com/vaadin/web-components/pull/2172 has been merged.

vlukashov replied:
👍

vlukashov replied:
`vaadin-virtual-list` does not clear the content rendered by the old renderer when the renderer changes (see https://github.com/vaadin/web-components/issues/2235), so this workaround is still necessary here.

Commit:
chore: remove unnecessary workaround

Removing `_$litPart$` is no longer necessary here since vaadin/web-components#2172 was merged.

Commit:
Revert "chore: remove unnecessary workaround"

This reverts commit c43cda41
the workaround is still necessary as the original issues in not yet fixed for the vaadin-virtual-list component

Commit:
chore: update the linked issue in lit-renderer.ts

`vaadin-virtual-list` does not clear the content rendered by the old renderer when the renderer changes (see https://github.com/vaadin/web-components/issues/2235), so this workaround is still necessary here.

tomivirkki committed:
refactor: align the package with the existing renderers

tomivirkki committed:
Merge branch 'master' into lit-renderer

tomivirkki committed:
chore: bump version numbers

tomivirkki committed:
chore: fix a linter issue

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/flow-components?pullRequest=1838) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/flow-components?pullRequest=1838) before we can accept your contribution.<br/><hr/>**Viktor Lukashov** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/flow-components?pullRequest=1838) it.</sub>

tomivirkki committed:
Merge remote-tracking branch 'origin/master' into lit-renderer

# Conflicts:
#	vaadin-combo-box-flow-parent/vaadin-combo-box-flow/pom.xml
